File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/presentation/console/ConsoleApp.java
package com.simple.ecommerce.presentation.console;

import com.simple.ecommerce.application.exception.CategoryNotFoundException;
import com.simple.ecommerce.application.exception.EmptyCartException;
import com.simple.ecommerce.application.exception.InvalidPromotionException;
import com.simple.ecommerce.application.exception.ProductNotFoundException;
import com.simple.ecommerce.application.exception.ProductUnavailableException;
import com.simple.ecommerce.application.service.CatalogService;
import com.simple.ecommerce.application.service.ShoppingCartService;
import com.simple.ecommerce.domain.entity.Category;
import com.simple.ecommerce.domain.entity.Product;
import com.simple.ecommerce.domain.repository.ProductRepository;
import com.simple.ecommerce.infrastructure.persistence.InMemoryProductRepository;
import com.simple.ecommerce.infrastructure.promotion.CheapestProductPromotion;
import com.simple.ecommerce.infrastructure.promotion.PercentagePromotion;
import com.simple.ecommerce.infrastructure.promotion.SecondProductHalfPricePromotion;

import java.util.List;
import java.util.Map;

/**
 * Console application for the eCommerce system.
 */
public class ConsoleApp {
    public static void main(String[] args) {
        System.out.println("Welcome to Simple eCommerce System");
        System.out.println("==================================\n");
        
        // Initialize repositories and services
        ProductRepository productRepository = new InMemoryProductRepository();
        CatalogService catalogService = new CatalogService(productRepository);
        ShoppingCartService cartService = new ShoppingCartService();
        
        // Initialize catalog with sample data
        initializeCatalog(catalogService);
        
        // Display all products alphabetically
        System.out.println("All products (sorted alphabetically):");
        List<Product> sortedProducts = catalogService.getProductsSortedAlphabetically();
        sortedProducts.forEach(product -> 
            System.out.println(" - " + product.getName() + " ($" + product.getPrice() + ")"));
        
        // Display products by category
        Category electronics = new Category("Electronics");
        System.out.println("\nElectronics products (sorted by price):");
        List<Product> electronicProducts = catalogService.getAvailableProductsByCategory(electronics);
        electronicProducts.forEach(product -> 
            System.out.println(" - " + product.getName() + " ($" + product.getPrice() + ")"));
        
        // Demonstrate shopping cart
        demonstrateShoppingCart(sortedProducts, cartService);
        
        // Demonstrate promotions
        demonstratePromotions(cartService);
        
        // Demonstrate exception handling
        System.out.println("\n=== Exception Handling Demonstration ===");
        demonstrateExceptionHandling(catalogService, cartService);
    }
    
    /**
     * Initializes the catalog with sample products.
     *
     * @param catalogService the catalog service to use
     */
    private static void initializeCatalog(CatalogService catalogService) {
        // Create categories
        Category electronics = new Category("Electronics");
        Category books = new Category("Books");
        Category clothing = new Category("Clothing");
        
        // Electronics
        catalogService.addProduct(new Product("Laptop", 1200.0, electronics));
        catalogService.addProduct(new Product("Smartphone", 800.0, electronics));
        catalogService.addProduct(new Product("Headphones", 150.0, electronics));
        catalogService.addProduct(new Product("Tablet", 500.0, electronics));
        catalogService.addProduct(new Product("Smart Watch", 250.0, electronics));
        
        // Books
        catalogService.addProduct(new Product("Java Programming", 45.0, books));
        catalogService.addProduct(new Product("Clean Code", 35.0, books));
        catalogService.addProduct(new Product("Design Patterns", 40.0, books));
        catalogService.addProduct(new Product("Algorithms", 50.0, books));
        
        // Clothing
        catalogService.addProduct(new Product("T-Shirt", 25.0, clothing));
        catalogService.addProduct(new Product("Jeans", 60.0, clothing));
        catalogService.addProduct(new Product("Hoodie", 45.0, clothing));
        
        // Mark one product as unavailable
        Product unavailableProduct = new Product("Out of Stock Item", 99.0, electronics);
        unavailableProduct.setAvailable(false);
        catalogService.addProduct(unavailableProduct);
    }
    
    /**
     * Demonstrates adding products to the shopping cart.
     *
     * @param products the available products
     * @param cartService the shopping cart service to use
     */
    private static void demonstrateShoppingCart(List<Product> products, ShoppingCartService cartService) {
        // Find products by name
        Product laptop = findProductByName(products, "Laptop");
        Product smartphone = findProductByName(products, "Smartphone");
        Product headphones = findProductByName(products, "Headphones");
        
        // Add products to cart
        cartService.addProduct(laptop);
        cartService.addProduct(smartphone);
        cartService.addProduct(headphones);
        cartService.addProduct(headphones); // Adding a second pair of headphones
        
        // Display cart contents
        System.out.println("\nShopping Cart Contents:");
        Map<Product, Integer> cartContents = cartService.getCartContents();
        cartContents.forEach((product, quantity) -> 
            System.out.println(" - " + product.getName() + " x" + quantity + 
                              " ($" + product.getPrice() + " each)"));
        
        // Calculate and display total price without promotions
        System.out.println("\nTotal Price: $" + String.format("%.2f", cartService.calculateCartPrice()));
    }
    
    /**
     * Demonstrates applying different promotions to the shopping cart.
     *
     * @param cartService the shopping cart service to use
     */
    private static void demonstratePromotions(ShoppingCartService cartService) {
        // 10% off everything promotion
        PercentagePromotion tenPercentOff = new PercentagePromotion("10PERCENTOFF");
        cartService.activatePromotion(tenPercentOff);
        System.out.println("\nWith 10% off promotion:");
        System.out.println("Total Price: $" + String.format("%.2f", cartService.calculateCartPrice()));
        
        // Every 3rd item costs 1 PLN promotion
        CheapestProductPromotion everyThirdCheapPromotion = new CheapestProductPromotion("EVERY3RD1PLN");
        cartService.activatePromotion(everyThirdCheapPromotion);
        System.out.println("\nWith every 3rd item for 1 PLN promotion:");
        System.out.println("Total Price: $" + String.format("%.2f", cartService.calculateCartPrice()));
        
        // Second item half price promotion
        SecondProductHalfPricePromotion secondHalfPricePromotion = new SecondProductHalfPricePromotion("SECONDHALFPRICE");
        cartService.activatePromotion(secondHalfPricePromotion);
        System.out.println("\nWith second item half price promotion:");
        System.out.println("Total Price: $" + String.format("%.2f", cartService.calculateCartPrice()));
    }
    
    /**
     * Demonstrates exception handling for various scenarios.
     *
     * @param catalogService the catalog service to use
     * @param cartService the shopping cart service to use
     */
    private static void demonstrateExceptionHandling(CatalogService catalogService, ShoppingCartService cartService) {
        // 1. ProductNotFoundException
        System.out.println("\n1. ProductNotFoundException handling:");
        try {
            System.out.println("  Trying to find non-existent product 'Gaming Console'...");
            Product nonExistentProduct = catalogService.findProductByName("Gaming Console");
            System.out.println("  Found product: " + nonExistentProduct.getName());
        } catch (ProductNotFoundException e) {
            System.out.println("  Error: " + e.getMessage());
        }
        
        // 2. ProductUnavailableException
        System.out.println("\n2. ProductUnavailableException handling:");
        try {
            System.out.println("  Trying to add unavailable product to cart...");
            Product unavailableProduct = findProductByName(catalogService.getAllProducts(), "Out of Stock Item");
            cartService.addProduct(unavailableProduct);
            System.out.println("  Product added successfully");
        } catch (ProductUnavailableException e) {
            System.out.println("  Error: " + e.getMessage());
            System.out.println("  Unavailable product: " + e.getProduct().getName());
        }
        
        // 3. CategoryNotFoundException
        System.out.println("\n3. CategoryNotFoundException handling:");
        try {
            System.out.println("  Trying to get products with null category...");
            catalogService.getAvailableProductsByCategory(null);
            System.out.println("  Products retrieved successfully");
        } catch (CategoryNotFoundException e) {
            System.out.println("  Error: " + e.getMessage());
        }
        
        // 4. InvalidPromotionException
        System.out.println("\n4. InvalidPromotionException handling:");
        try {
            System.out.println("  Trying to create promotion with invalid discount percentage...");
            PercentagePromotion invalidPromotion = new PercentagePromotion("INVALID_PROMO", 110.0);
            System.out.println("  Promotion created successfully: " + invalidPromotion.getPromotionCode());
        } catch (InvalidPromotionException e) {
            System.out.println("  Error: " + e.getMessage());
        }
        
        // 5. EmptyCartException
        System.out.println("\n5. EmptyCartException handling:");
        try {
            System.out.println("  Trying to calculate price with empty cart...");
            // Create a new cart that is empty
            ShoppingCartService emptyCart = new ShoppingCartService();
            emptyCart.calculateCartPrice();
            System.out.println("  Price calculated successfully");
        } catch (EmptyCartException e) {
            System.out.println("  Error: " + e.getMessage());
        }
        
        // 6. Demonstrate proper exception recovery
        System.out.println("\n6. Exception recovery demonstration:");
        ShoppingCartService recoveryCart = new ShoppingCartService();
        
        // Try to calculate price with empty cart
        try {
            System.out.println("  Trying to calculate price with empty cart...");
            recoveryCart.calculateCartPrice();
        } catch (EmptyCartException e) {
            System.out.println("  Error: " + e.getMessage());
            System.out.println("  Recovering by adding a product to cart...");
            
            // Add a product and try again
            Product laptop = findProductByName(catalogService.getAllProducts(), "Laptop");
            recoveryCart.addProduct(laptop);
            
            try {
                double price = recoveryCart.calculateCartPrice();
                System.out.println("  Recovery successful! Cart price: $" + String.format("%.2f", price));
            } catch (Exception recoveryException) {
                System.out.println("  Recovery failed: " + recoveryException.getMessage());
            }
        }
        
        System.out.println("\nException handling demonstration completed.");
    }
    
    /**
     * Helper method to find a product by name from a list of products.
     *
     * @param products the list of products to search in
     * @param name the name of the product to find
     * @return the found product or null if not found
     */
    private static Product findProductByName(List<Product> products, String name) {
        return products.stream()
                .filter(p -> p.getName().equals(name))
                .findFirst()
                .orElse(null);
    }
} 

File: Projects/Testtrack/p911 - Simple eCommerce backend/presentation/console/console_app.py
from typing import List, Optional
from application import CategoryNotFoundException, EmptyCartException, InvalidPromotionException, ProductNotFoundException, ProductUnavailableException, CatalogService, ShoppingCartService
from domain import Category, Product
from infrastructure import InMemoryProductRepository, CheapestProductPromotion, PercentagePromotion, SecondProductHalfPricePromotion


def main():
    print("Welcome to Simple eCommerce System")
    print("==================================\n")

    product_repository = InMemoryProductRepository()
    catalog_service = CatalogService(product_repository)
    cart_service = ShoppingCartService()

    initialize_catalog(catalog_service)

    print("All products (sorted alphabetically):")
    sorted_products = catalog_service.get_products_sorted_alphabetically()
    for product in sorted_products:
        print(f" - {product.get_name()} (${product.get_price()})")

    electronics = Category("Electronics")
    print("\nElectronics products (sorted by price):")
    electronic_products = catalog_service.get_available_products_by_category(
        electronics)
    for product in electronic_products:
        print(f" - {product.get_name()} (${product.get_price()})")

    demonstrate_shopping_cart(sorted_products, cart_service)
    demonstrate_promotions(cart_service)

    print("\n=== Exception Handling Demonstration ===")
    demonstrate_exception_handling(catalog_service, cart_service)


def initialize_catalog(catalog_service: CatalogService) -> None:
    electronics = Category("Electronics")
    books = Category("Books")
    clothing = Category("Clothing")

    catalog_service.add_product(Product("Laptop", 1200.0, electronics))
    catalog_service.add_product(Product("Smartphone", 800.0, electronics))
    catalog_service.add_product(Product("Headphones", 150.0, electronics))
    catalog_service.add_product(Product("Tablet", 500.0, electronics))
    catalog_service.add_product(Product("Smart Watch", 250.0, electronics))

    catalog_service.add_product(Product("Java Programming", 45.0, books))
    catalog_service.add_product(Product("Clean Code", 35.0, books))
    catalog_service.add_product(Product("Design Patterns", 40.0, books))
    catalog_service.add_product(Product("Algorithms", 50.0, books))

    catalog_service.add_product(Product("T-Shirt", 25.0, clothing))
    catalog_service.add_product(Product("Jeans", 60.0, clothing))
    catalog_service.add_product(Product("Hoodie", 45.0, clothing))

    unavailable_product = Product("Out of Stock Item", 99.0, electronics)
    unavailable_product.set_available(False)
    catalog_service.add_product(unavailable_product)


def demonstrate_shopping_cart(products: List[Product], cart_service: ShoppingCartService) -> None:
    laptop = find_product_by_name(products, "Laptop")
    smartphone = find_product_by_name(products, "Smartphone")
    headphones = find_product_by_name(products, "Headphones")

    cart_service.add_product(laptop)
    cart_service.add_product(smartphone)
    cart_service.add_product(headphones)
    cart_service.add_product(headphones)

    print("\nShopping Cart Contents:")
    cart_contents = cart_service.get_cart_contents()
    for product, quantity in cart_contents.items():
        print(f" - {product.get_name()} x{quantity} (${product.get_price()} each)")

    print(f"\nTotal Price: ${cart_service.calculate_cart_price():.2f}")


def demonstrate_promotions(cart_service: ShoppingCartService) -> None:
    ten_percent_off = PercentagePromotion("10PERCENTOFF")
    cart_service.activate_promotion(ten_percent_off)
    print("\nWith 10% off promotion:")
    print(f"Total Price: ${cart_service.calculate_cart_price():.2f}")

    every_third_cheap_promotion = CheapestProductPromotion("EVERY3RD1PLN")
    cart_service.activate_promotion(every_third_cheap_promotion)
    print("\nWith every 3rd item for 1 PLN promotion:")
    print(f"Total Price: ${cart_service.calculate_cart_price():.2f}")

    second_half_price_promotion = SecondProductHalfPricePromotion(
        "SECONDHALFPRICE")
    cart_service.activate_promotion(second_half_price_promotion)
    print("\nWith second item half price promotion:")
    print(f"Total Price: ${cart_service.calculate_cart_price():.2f}")


def demonstrate_exception_handling(catalog_service: CatalogService, cart_service: ShoppingCartService) -> None:
    print("\n1. ProductNotFoundException handling:")
    try:
        print("  Trying to find non-existent product 'Gaming Console'...")
        non_existent_product = catalog_service.find_product_by_name(
            "Gaming Console")
        print(f"  Found product: {non_existent_product.get_name()}")
    except ProductNotFoundException as e:
        print(f"  Error: {e}")

    print("\n2. ProductUnavailableException handling:")
    try:
        print("  Trying to add unavailable product to cart...")
        unavailable_product = find_product_by_name(
            catalog_service.get_all_products(), "Out of Stock Item")
        cart_service.add_product(unavailable_product)
        print("  Product added successfully")
    except ProductUnavailableException as e:
        print(f"  Error: {e}")
        print(f"  Unavailable product: {e.get_product().get_name()}")

    print("\n3. CategoryNotFoundException handling:")
    try:
        print("  Trying to get products with null category...")
        catalog_service.get_available_products_by_category(None)
        print("  Products retrieved successfully")
    except CategoryNotFoundException as e:
        print(f"  Error: {e}")

    print("\n4. InvalidPromotionException handling:")
    try:
        print("  Trying to create promotion with invalid discount percentage...")
        invalid_promotion = PercentagePromotion("INVALID_PROMO", 110.0)
        print(
            f"  Promotion created successfully: {invalid_promotion.get_promotion_code()}")
    except InvalidPromotionException as e:
        print(f"  Error: {e}")

    print("\n5. EmptyCartException handling:")
    try:
        print("  Trying to calculate price with empty cart...")
        empty_cart = ShoppingCartService()
        empty_cart.calculate_cart_price()
        print("  Price calculated successfully")
    except EmptyCartException as e:
        print(f"  Error: {e}")

    print("\n6. Exception recovery demonstration:")
    recovery_cart = ShoppingCartService()

    try:
        print("  Trying to calculate price with empty cart...")
        recovery_cart.calculate_cart_price()
    except EmptyCartException as e:
        print(f"  Error: {e}")
        print("  Recovering by adding a product to cart...")

        laptop = find_product_by_name(
            catalog_service.get_all_products(), "Laptop")
        recovery_cart.add_product(laptop)

        try:
            price = recovery_cart.calculate_cart_price()
            print(f"  Recovery successful! Cart price: ${price:.2f}")
        except Exception as recovery_exception:
            print(f"  Recovery failed: {recovery_exception}")

    print("\nException handling demonstration completed.")


def find_product_by_name(products: List[Product], name: str) -> Optional[Product]:
    for product in products:
        if product.get_name() == name:
            return product
    return None


if __name__ == "__main__":
    main()
